src/features/UpdatePowerShell.ts

```diff
@@ -0,0 +1,146 @@
+        let releaseJson: any;
+        if (preview) {
+            // This gets all releases and the first one is the latest prerelease if
+            // there is a prerelease version.
+            releaseJson = (await fetch(PowerShellGitHubPrereleasesUrl)
+                .then((res) => res.json())).find((release: any) => release.prerelease);
+        } else {
+            releaseJson = await fetch(PowerShellGitHubReleasesUrl)
+                .then((res) => res.json());
+        }
+
+        return new GitHubReleaseInformation(
+            releaseJson.tag_name, releaseJson.assets);
+    }
+
+    public version: SemVer;
+    public isPreview: boolean = false;
+    public assets: any[];
+
+    public constructor(version: string | SemVer, assets: any[] = []) {
+        this.version = parse(version);
+
+        if (prerelease(this.version)) {
+            this.isPreview = true;
+        }
+
+        this.assets = assets;
+    }
+}
+
+interface IUpdateMessageItem extends MessageItem {
+    id: number;
+}
+
+export async function InvokePowerShellUpdateCheck(
+    languageServerClient: LanguageClient,
+    localVersion: SemVer,
+    arch: string,
+    release: GitHubReleaseInformation) {
+    const options: IUpdateMessageItem[] = [
+        {
+            id: 0,
+            title: "Yes!",
+        },
+        {
+            id: 1,
+            title: "Not now",
+        },
+        {
+            id: 2,
+            title: "Don't show me this notification again...",
+        },
+    ];
+
+    // If our local version is up-to-date, we can return early.
+    if (compare(localVersion, release.version) >= 0) {
+        return;
+    }
+
+    const commonText: string = `You have an old version of PowerShell (${
+        localVersion.raw
+    }). The current latest release is ${
+        release.version.raw
+    }.`;
+
+    if (process.platform === "linux") {
+        await window.showInformationMessage(
+            `${commonText} We recommend updating to the latest version.`);
+        return;
+    }
+
+    const isMacOS: boolean = process.platform === "darwin";
+    const result = await window.showInformationMessage(
+        `${commonText} Would you like to update the version? ${
+            isMacOS ? "(Homebrew is required on macOS)" : ""
+        }`, ...options);
+
+    // If the user cancels the notification.
+    if (!result) { return; }
+
+    // Yes choice.
+    switch (result.id) {
+        // Yes choice.
+        case 0:
+            let script: string;
+            if (process.platform === "win32") {
+                const msiMatcher = arch === "x86" ?
+                    "win-x86.msi" : "win-x64.msi";
+
+                const assetUrl = release.assets.filter((asset: any) =>
+                    asset.name.indexOf(msiMatcher) >= 0)[0].url;
+
+                // Grab MSI and run it.
+                // tslint:disable-next-line: max-line-length
+                script = `
+$randomFileName = [System.IO.Path]::GetRandomFileName()
```

Comment:
Cool, I did not know that API existed (even on older .Net versions). Also `+1` for not using the bad GetTempPath API